[proxy] add analytics endpoint

akosyakov opened the pull request:
## Description
<!-- Describe your changes in detail -->

[Internal context](https://gitpod.slack.com/archives/C047WL9JBLG/p1680019749050709)

- This PR introduces a new `/analytics` endpoint for collecting analytics from different clients in a centralized manner. The endpoint utilizes basic authentication from segment to determine whether to proxy to trusted or untrusted sources.
- In preview environments, Gitpod components (server, ws-manager-bridge, ws-proxy) are configured to use this endpoint with a trusted key, and the same configuration will be applied to production environments in a subsequent PR.
- In workspace services (supervisor and Gitpod CLI) are also reconfigured to use this endpoint but without a key in preview environments (i.e., untrusted), and the same configuration to be applied to the production environment in a subsequent PR.
- After deployment, local clients like VS Code Desktop will also be updated to use this endpoint.
- This PR as well fixes analytics installer options for GH actions.

## Related Issue(s)
<!-- List the issue(s) this PR solves -->
Fixes #

## How to test
<!-- Provide steps to test this PR -->

- Start a new workspace and check that analytics is configured now based on segment endpoint not key.
<img width="1060" alt="Screenshot 2023-04-12 at 21 39 29" src="https://user-images.githubusercontent.com/3082655/231568577-7095bfa6-0c35-4e11-8a1d-ad89cf5cf20b.png">

- Run gp tasks list command and then check staging trusted that gp_command event is registered there.
- Check that events from other Gitpod components are registered as well, i.e. workspace started and stopped, and so on.

## Release Notes
<!--
  Add entries for the CHANGELOG.md or "NONE" if there aren't any user facing changes.
  Each line becomes a separate entry.
  Format: [!<optional for breaking>] <description>
  Example: !basic auth is no longer supported
  See https://www.notion.so/gitpod/Release-Notes-513a74fdd23b4cb1b3b3aefb1d34a3e0
-->
```release-note
NONE
```

## Documentation
<!--
Does this PR require updates to the documentation at www.gitpod.io/docs?
* Yes
  * 1. Please create a docs issue: https://github.com/gitpod-io/website/issues/new?labels=documentation&template=DOCS-NEW-FEATURE.yml&title=%5BDocs+-+New+Feature%5D%3A+%3Cyour+feature+name+here%3E
  * 2. Paste the link to the docs issue below this comment
* No
  * Are you sure? If so, nothing to do here.
-->

## Build Options:

- [ ] /werft with-werft
      Run the build with werft instead of GHA
- [ ] leeway-no-cache
- [ ] /werft no-test
      Run Leeway with `--dont-test`

<details>
<summary>Publish Options</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer Options</summary>

- [ ] with-dedicated-emulation
- [ ] with-ws-manager-mk2
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

#### Preview Environment Options:
- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [ ] /werft with-large-vm
- [ ] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [ ] with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`
- [x] analytics=segment

/hold

roboquat commented:
@akosyakov: Adding the "do-not-merge/release-note-label-needed" label because no release-note block was detected, please follow our [release note process](https://git.k8s.io/community/contributors/guide/release-notes.md) to remove it.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

filiptronicek commented:
Would rate limiting be in-scope for this PR? Maybe it would be useful if we are going to implement the reverse proxy anyway

akosyakov commented:
> Would rate limiting be in-scope for this PR? Maybe it would be useful if we are going to implement the reverse proxy anyway

Was not planing to till we see actual abuse.

akosyakov commented:
/gh run recreate-vm

akosyakov commented on components/proxy/plugins/analytics/analytics.go:
```diff
@@ -0,0 +1,125 @@
+		ExpectContinueTimeout: 1 * time.Second,
+	}
+
+	a.reverseProxy = http.StripPrefix("/analytics", reverseProxy)
```
probably it is better to add `/analytics/segment` route that we can add other routes later if necessary, not only proxy to segment

easyCZ commented:
@akosyakov given our discussion yesterday, is this something you're looking to land as is? If not, could we move to draft until it's eligible for review?

akosyakov commented:
> @akosyakov given our discussion yesterday, is this something you're looking to land as is? If not, could we move to draft until it's eligible for review?

Refactored it since yesterday, but I will need your input. Now `/analytics` endpoint accept from all clients, and decide whether to delegate to trusted or untrusted source based on segment basic auth, i.e. if client is using trusted key then it goes to trusted, if no key or untrusted then to untrusted.

easyCZ commented on components/common-go/analytics/analytics.go:
```diff
@@ -45,7 +45,14 @@ func NewFromEnvironment() Writer {
+			Endpoint: os.Getenv("GITPOD_ANALYTICS_SEGMENT_ENDPOINT"),
+		})
+		if err != nil {
+			log.WithError(err).Error("cannot create segment client")
```
How critical are analytics? Do we want to exit/propagate error if we can't set it up?

akosyakov replied:
I cannot answer for everybody. In workspace it is not super critical to interrupt user flow, i.e. if we don't know about usage of gp cli it is not the end of the world. Not sure about other components like dashboard.

easyCZ commented on components/proxy/plugins/analytics/analytics.go:
```diff
@@ -0,0 +1,143 @@
+	}
+}
+
+func (a *Analytics) ServeHTTP(w http.ResponseWriter, r *http.Request, next caddyhttp.Handler) error {
+	segmentKey, _, ok := r.BasicAuth()
+	if !ok {
+		return next.ServeHTTP(w, r)
+	}
+	if a.trustedProxy != nil && segmentKey == a.trustedProxy.segmentKey {
+		a.trustedProxy.ServeHTTP(w, r)
+		return nil
+	}
+	if a.untrustedProxy != nil && (segmentKey == "" || segmentKey == a.untrustedProxy.segmentKey) {
+		if segmentKey == "" {
+			r.SetBasicAuth(a.untrustedProxy.segmentKey, "")
+		}
+		a.untrustedProxy.ServeHTTP(w, r)
+		return nil
+	}
+	return next.ServeHTTP(w, r)
+}
```
Would be nice to have a test for this

easyCZ commented on install/installer/pkg/components/proxy/deployment.go:
```diff
@@ -258,6 +264,12 @@ func deployment(ctx *common.RenderContext) ([]runtime.Object, error) {
 								}, {
 									Name:  "WORKSPACE_HANDLER_FILE",
 									Value: strings.ToLower(string(ctx.Config.Kind)),
+								}, {
+									Name:  "ANALYTICS_PLUGIN_TRUSTED_SEGMENT_KEY",
+									Value: trustedSegmentKey,
+								}, {
+									Name:  "ANALYTICS_PLUGIN_UNTRUSTED_SEGMENT_KEY",
+									Value: untrustedSegmentKey,
 								}},
```
Can we pull this into something like `common.AnalyticsEnv` similar to other re-usable parts?

akosyakov replied:
Maybe I misunderstand, but I think this PR already adds such config: https://github.com/gitpod-io/gitpod/pull/17162/files#diff-590a17333c6831c6ff532edbab0b818ee3c49fb033d163722552bbbfafd31a8cR466

easyCZ replied:
You're right, sorry. Missed that.

easyCZ commented on install/installer/pkg/config/v1/config.go:
```diff
@@ -180,8 +180,9 @@ type Observability struct {
-	SegmentKey string `json:"segmentKey"`
-	Writer     string `json:"writer"`
+	SegmentKey      string `json:"segmentKey"`
+	Writer          string `json:"writer"`
```
Out of curiosity, what does Writer mean?

akosyakov replied:
i think here it is none, log or segment

easyCZ commented on components/proxy/plugins/analytics/analytics.go:
```diff
@@ -0,0 +1,143 @@
+		a.trustedProxy.ServeHTTP(w, r)
+		return nil
+	}
+	if a.untrustedProxy != nil && (segmentKey == "" || segmentKey == a.untrustedProxy.segmentKey) {
```
```suggestion
	shouldProxytoUntrusted := a.untrustedProxy != nil && (segmentKey == "" || segmentKey == a.untrustedProxy.segmentKey)
	if  shouldProxyToUntrusted {
```
Mostly for ease of legibility as the expression can be a mouthful

easyCZ commented on components/proxy/plugins/analytics/analytics.go:
```diff
@@ -0,0 +1,143 @@
+	return nil
+}
+
+func newSegmentProxy(segmentEndponit *url.URL, errorLog *log.Logger, segmentKey string) *segmentProxy {
```
```suggestion
func newSegmentProxy(segmentEndpoint *url.URL, errorLog *log.Logger, segmentKey string) *segmentProxy {
```

easyCZ reviewed (commented):
In general looks good

akosyakov commented:
I've addressed comments and added tests. I will test more manually later. @mads-hartmann maybe you can have a look whether it is something what you think would help you with dedicated

mads-hartmann reviewed (commented):
I have two questions related to dedicated

1. Initially we'll want the analytics endpoint to just drop all events in Dedicated. I couldn't quite tell if it's currently possible to have the analytics caddy plugin drop events right now or we'd need to modify it to achieve that.
2. In the near future we'll want to (1) sanitise events (2) send them to a Firehose Delivery Stream instead of Segment. It looks like Caddy plugins are powerful enough that we'll be able to do that, is that your understanding too?

akosyakov commented:
> Initially we'll want the analytics endpoint to just drop all events in Dedicated. I couldn't quite tell if it's currently possible to have the analytics caddy plugin drop events right now or we'd need to modify it to achieve that.

if you don't configure any segment keys, analytics plugin will drop everything.

> In the near future we'll want to (1) sanitise events (2) send them to a Firehose Delivery Stream instead of Segment. It looks like Caddy plugins are powerful enough that we'll be able to do that, is that your understanding too?

yes, it is go program, you have access to request object to modify the body before forwarding somewhere

mads-hartmann reviewed (approved):
Based on [this comment](https://github.com/gitpod-io/gitpod/pull/17162#issuecomment-1508211032) this PR looks good to me! 🛹

akosyakov commented:
I added tests for cases when keys are not configured.

akosyakov commented:
/gh run recreate-vm=true
